File: server/branding.py
"""branding.py — the app's own NAME and MARK, moved out of the JSX and into the
admin panel.

THE PROBLEM THIS SOLVES, in one sentence: the product name was typed into eight
files and the mark was drawn in two more, so renaming the app meant a developer,
an editor and a redeploy — and the drawn mark could never be the customer's own
logo at all.

⚠ **ONE ROW, NOT A LIST.** There is exactly one product being named, so this
store holds a SINGLE document under `_DOC_ID` and every reader asks for that same
one. It is the only store in `server/` shaped that way, and it is deliberate: a
"brandings" collection with one row in it invites a second row nobody can
explain, and a `GET /public/branding` that has to pick which one is live.

⚠ **THE READ IS PUBLIC AND UNAUTHENTICATED, BECAUSE THE NAME IS ON THE LANDING
PAGE.** The sign-in card, the marketing page and a shared storyboard link are all
seen by somebody with no account, and every one of them prints the name and draws
the mark. Same reasoning as `/public/workflows` and `GET /billing/tiers`: what
you are CALLED is public by nature. Nothing here leaks anything else — the
payload is a name, a URL and a stamp.

⚠ **THE STAMP IS THE CACHE.** The logo's URL carries the id of the file it is
serving (`/public/branding/logo/{stamp}`), so a NEW upload is a NEW address —
every browser, every proxy and every already-open tab picks it up the moment the
name call answers, and nobody has to reason about cache headers. That is why the
id is regenerated on every upload rather than the bytes being written over the
same path: overwriting one URL is exactly how a logo change fails to appear for
the one person who most needed to see it.

⚠ **TWO LOGO SLOTS, ONE PER THEME, AND THAT IS NOT A LUXURY.** A logo is a
FLAT PICTURE — unlike the drawn mark below it, which is painted in `currentColor`
and re-colours itself for free. The first white wordmark uploaded here vanished
into the light theme completely: *"jab mai dark mode mai hun to mera logo dikh
raha hai magar jab light mode mai karta hun to mera logo white mai merge ho raha
hai."* So there is a `dark` slot and a `light` slot, and the app picks by theme.

⚠ **EITHER SLOT FILLS IN FOR THE OTHER** (`resolve_slot`). One upload is a
complete, working answer — the commonest logo is a full-colour mark that reads on
both grounds, and forcing somebody to upload the same file twice to get started
would be a worse product than the one-slot version this replaces. The second slot
is there for when one file genuinely cannot do both.

⚠ **IT FAILS BACK TO THE BUILT-IN, NEVER TO NOTHING.** An unreachable store, a
missing file, a half-finished upload — every one of them answers with the
compiled-in name and no logo, which is the app exactly as it shipped. A brand
call that fails must never produce a screen with no title on it.

Backend follows the USER STORE (`API_USER_STORE`), like every other panel store.
"""

import json
import logging
import os
import re
import threading
import time
import uuid
from datetime import datetime, timezone
from pathlib import Path

from fastapi import APIRouter, HTTPException
from fastapi.responses import FileResponse

from . import config

logger = logging.getLogger(__name__)

router = APIRouter(tags=["branding"])

_lock = threading.Lock()

# The single document's id, in both stores.
_DOC_ID = "app"

# ⚠ THE COMPILED-IN NAME, AND THE ONE PLACE IT IS STILL WRITTEN DOWN. Every
# screen used to carry its own copy of this string; they all read the store now,
# and the store falls back to here. Changing the shipped default is a code
# change — changing what customers SEE is a click in the panel.
DEFAULT_NAME = "Aniwala AI Studio"

# A name has to fit the sidebar, which ellipsises (`.sb-brand-name`), and the
# admin top bar, which does too. 40 is measured against the 280px rail at its
# real font — see `tests/brand_landing_check.py`, which fails if the rail trims.
NAME_MAX_CHARS = 40

# What the panel may write. Anything else in a PATCH body is ignored rather than
# rejected, the same way `features.save_feature` treats its own EDITABLE set.
EDITABLE = frozenset({"name"})

# The logo, once normalised. Kept small because EVERY visitor downloads it on
# EVERY cold page load, including the marketing page — a 4000px master would be
# a megabyte on the front door.
LOGO_MAX_PX = 512
ALLOWED_LOGO_TYPES = ("image/png", "image/jpeg", "image/webp")

# ---------------------------------------------------------------------------
# The two slots
# ---------------------------------------------------------------------------
# ⚠ THE DARK SLOT IS STORED UNDER THE UN-SUFFIXED `logo_id`, AND THAT IS NOT
# UNTIDINESS — it is the field every deployment that uploaded a logo before the
# split already has. Renaming it to `logo_dark_id` would orphan those files: the
# app would come back up wearing the built-in mark and the person who uploaded
# theirs would have no idea why. `dark` is also the right one to inherit it,
# because dark is this app's default theme and is what that upload was made for.
SLOT_DARK = "dark"
SLOT_LIGHT = "light"
SLOTS = (SLOT_DARK, SLOT_LIGHT)
_SLOT_FIELD = {SLOT_DARK: "logo_id", SLOT_LIGHT: "logo_light_id"}
# The slot each one borrows from when it is empty. See `resolve_slot`.
_SLOT_FALLBACK = {SLOT_DARK: SLOT_LIGHT, SLOT_LIGHT: SLOT_DARK}


def slot_field(slot: str) -> str:
    """`"light"` → `"logo_light_id"`. Raises ValueError on anything else."""
    try:
        return _SLOT_FIELD[slot]
    except KeyError:
        raise ValueError(f"Unknown logo slot: {slot!r}. Use one of {', '.join(SLOTS)}.")


def resolve_slot(row: dict, slot: str) -> str:
    """The logo id a theme should actually draw — its own, or the other one.

    ⚠ THE FALLBACK IS THE WHOLE REASON ONE UPLOAD IS ENOUGH. A deployment with a
    single full-colour mark uploads it once and both themes use it; only somebody
    whose mark genuinely cannot work on both grounds ever touches the second slot.
    """
    own = row.get(slot_field(slot)) or ""
    if own:
        return own
    return row.get(slot_field(_SLOT_FALLBACK[slot])) or ""


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat()


def _defaults() -> dict:
    return {
        "id": _DOC_ID,
        "name": DEFAULT_NAME,
        # "" in BOTH means: draw the built-in mark
        # (`client/src/components/Logo.jsx`). "" in one of them means: borrow the
        # other — see `resolve_slot`. ⚠ `logo_id` IS THE DARK SLOT; read the note
        # on `SLOT_DARK` before renaming it.
        "logo_id": "",
        "logo_light_id": "",
        "updated_at": None,
        "updated_by": None,
    }


def clean_name(name: str | None) -> str:
    """A typed name, made safe to print. Empty → the built-in default.

    ⚠ WHITESPACE IS COLLAPSED, NOT JUST TRIMMED. The field is a textarea (it has
    to wrap while being typed), so a pasted name can arrive with a newline in the
    middle of it — and a newline inside `.sb-brand-name`, which is
    `white-space: nowrap`, is an invisible character that silently widens the row.
    """
    clean = re.sub(r"\s+", " ", (name or "")).strip()
    return clean[:NAME_MAX_CHARS] or DEFAULT_NAME


# ===========================================================================
# Storage
# ===========================================================================
def _use_local() -> bool:
    return config.USER_STORE == "local"


def _local_path() -> Path:
    return Path(config.LOCAL_BRANDING_PATH)


def _local_load() -> dict:
    path = _local_path()
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
        return data if isinstance(data, dict) else {}
    except (OSError, json.JSONDecodeError):
        logger.warning("Local branding store at %s is unreadable — using defaults.", path)
        return {}


def _local_save(data: dict) -> None:
    _local_path().write_text(json.dumps(data, indent=2), encoding="utf-8")


_collection = None


def get_collection():
    """The branding collection, connecting on first use."""
    global _collection
    if _collection is not None:
        return _collection
    with _lock:
        if _collection is not None:
            return _collection
        from .mongo import get_db

        col = get_db()[config.BRANDING_COLLECTION]
        try:
            col.create_index("id", unique=True)
        except Exception as e:  # noqa: BLE001 — an index is an optimisation
            logger.warning("Could not index branding (%s).", e)
        _collection = col
        logger.info(
            "MongoDB branding store ready (db=%s, collection=%s)",
            config.MONGODB_DB, config.BRANDING_COLLECTION,
        )
        return _collection


# ===========================================================================
# The cache
# ===========================================================================
# ⚠ ONLY EVER REPLACED BY A SUCCESSFUL READ, so a database that goes away leaves
# the last known-good name on screen instead of reverting every visitor to the
# built-in one mid-session. Same rule as `features._cache`.
_cache: dict | None = None
_cache_at: float = 0.0
_cache_lock = threading.Lock()


def _bump() -> None:
    """Drop the cache so the next read goes to the store. Called after a write."""
    global _cache, _cache_at
    with _cache_lock:
        _cache = None
        _cache_at = 0.0


def get_branding(fresh: bool = False) -> dict:
    """The one branding row, defaults merged in. NEVER RAISES.

    A field added in a later release appears with its default on a document
    written before it existed, exactly as `all_features` merges its catalogue.
    """
    global _cache, _cache_at
    now = time.monotonic()
    if not fresh:
        with _cache_lock:
            if _cache is not None and now - _cache_at < config.BRANDING_CACHE_TTL_S:
                return _cache

    merged = _defaults()
    try:
        stored = _read_stored()
        merged.update({k: v for k, v in stored.items() if v is not None and k != "_id"})
        merged["id"] = _DOC_ID
        merged["name"] = clean_name(merged.get("name"))
    except Exception as e:  # noqa: BLE001 — fail back to the built-in; see the docstring
        with _cache_lock:
            if _cache is not None:
                logger.warning("Branding store unreachable (%s) — serving the last known good.", e)
                return _cache
        logger.warning("Branding store unreachable (%s) — serving the built-in name.", e)
        return merged

    # ⚠ A LOGO ID WHOSE FILE IS GONE IS NOT A LOGO. The document and the file are
    # two different stores and they can drift — a restored database, a wiped
    # uploads volume — and the failure mode without this check is every screen in
    # the app drawing a broken-image icon where its mark should be. Checked PER
    # SLOT, so losing the light-mode file leaves the dark one working and the
    # light theme simply borrows it.
    for _slot in SLOTS:
        field = slot_field(_slot)
        if merged.get(field) and not os.path.isfile(logo_path(merged[field])):
            logger.warning("Branding logo %s is recorded but missing on disk.", merged[field])
            merged[field] = ""

    with _cache_lock:
        _cache = merged
        _cache_at = now
    return merged


def _read_stored() -> dict:
    if _use_local():
        with _lock:
            return _local_load()
    return get_collection().find_one({"id": _DOC_ID}) or {}


def save_branding(fields: dict, actor: str = "") -> dict:
    """Write the panel's changes onto the one row. Returns the merged result."""
    clean = {k: v for k, v in (fields or {}).items() if k in EDITABLE}
    if "name" in clean:
        clean["name"] = clean_name(clean["name"])
    _write(clean, actor)
    return get_branding(fresh=True)


def _write(clean: dict, actor: str) -> None:
    """The store half of a save. `clean` is already validated."""
    clean = dict(clean)
    clean["id"] = _DOC_ID
    clean["updated_at"] = _now_iso()
    clean["updated_by"] = (actor or "").strip().lower() or None

    if _use_local():
        with _lock:
            data = _local_load()
            data.update(clean)
            _local_save(data)
    else:
        get_collection().update_one({"id": _DOC_ID}, {"$set": clean}, upsert=True)
    _bump()


# ===========================================================================
# The logo file
# ===========================================================================
# ⚠ THE ID GOES INTO A FILE PATH AND INTO A URL, so it is checked against this
# rather than trusted. `save_logo` generates it and nothing else may, but the
# public route takes one straight off the wire.
_ID_RE = re.compile(r"^[a-f0-9]{12}$")


def logo_path(logo_id: str) -> str:
    """Where one logo's bytes live."""
    return os.path.join(config.BRANDING_DIR, f"{logo_id}.png")


def save_logo(png_bytes: bytes, slot: str = SLOT_DARK, actor: str = "") -> dict:
    """Store already-normalised PNG bytes in one slot. Returns the new row.

    ⚠ A NEW ID EVERY TIME, and the old file is deleted AFTER the document points
    at the new one — never before. Between those two writes the OLD logo is still
    the live one and is still on disk, so a crash mid-upload leaves the app
    wearing its previous mark rather than a broken image. See the module note on
    the stamp.
    """
    field = slot_field(slot)
    previous = get_branding(fresh=True).get(field) or ""
    logo_id = uuid.uuid4().hex[:12]
    os.makedirs(config.BRANDING_DIR, exist_ok=True)
    with open(logo_path(logo_id), "wb") as fh:
        fh.write(png_bytes)

    _write({field: logo_id}, actor)

    if previous and previous != logo_id:
        _delete_file(previous)
    return get_branding(fresh=True)


def clear_logo(slot: str = SLOT_DARK, actor: str = "") -> dict:
    """Empty one slot. Returns the new row.

    ⚠ CLEARING ONE SLOT IS NOT "NO LOGO" — the theme it belonged to falls back to
    the OTHER slot (`resolve_slot`), and only clearing BOTH puts the built-in
    drawn mark back. That is the honest behaviour and the panel says so, because
    "Remove" on one card silently blanking the other would be the worse surprise.
    """
    field = slot_field(slot)
    previous = get_branding(fresh=True).get(field) or ""
    _write({field: ""}, actor)
    if previous:
        _delete_file(previous)
    return get_branding(fresh=True)


def _delete_file(logo_id: str) -> None:
    """Remove a superseded logo.

    ⚠ NEVER RAISES. The document is already correct by the time this runs, so a
    file that will not delete is litter, not a fault — and turning it into one
    would fail an upload that has already succeeded.
    """
    if not _ID_RE.match(logo_id or ""):
        return
    try:
        os.remove(logo_path(logo_id))
    except OSError:
        pass


def normalise_logo(contents: bytes) -> bytes:
    """An uploaded image → the PNG that gets stored. Raises ValueError on junk.

    ⚠ RGBA, NOT RGB — the same rule as `POST /brand/logo` in `main.py`, and for
    the same reason: `.convert("RGB")` fills a transparent background with black,
    so a designer's transparent PNG would appear in the sidebar inside a hard
    black rectangle. This mark sits on `--panel` in the rail and on `--bg` on the
    landing page; it MUST keep its alpha.

    ⚠ AND IT IS DOWNSCALED, NOT REJECTED, when it is bigger than `LOGO_MAX_PX`.
    The person uploading has a print master and no way to resize it; refusing it
    would stop them branding the app at all. `thumbnail` keeps the aspect ratio,
    so a wide wordmark stays wide.
    """
    import io as _io

    from PIL import Image as PILImage

    try:
        img = PILImage.open(_io.BytesIO(contents))
        img.load()
        img = img.convert("RGBA")
    except Exception as e:  # noqa: BLE001 — bad/corrupt upload
        raise ValueError(f"Couldn't read that image: {e}")

    if max(img.size) > LOGO_MAX_PX:
        img.thumbnail((LOGO_MAX_PX, LOGO_MAX_PX), PILImage.LANCZOS)

    out = _io.BytesIO()
    img.save(out, "PNG")
    return out.getvalue()


def public_payload(row: dict | None = None) -> dict:
    """What a browser is told.

    ⚠ THE ONE SHAPE. The public route, the admin route and the tests all build
    the client's answer through here, so there is no second definition of what
    `logo_url` looks like.

    ⚠ **THE FALLBACK IS RESOLVED HERE, NOT IN THE BROWSER.** Both URLs come down
    already pointing at a real file, so a deployment with one upload sends the
    same address twice and the client needs no rule of its own. Putting
    "borrow the other slot" in the JSX would mean writing it again in the favicon
    code, and a rule written twice is a rule that disagrees with itself.
    """
    row = row or get_branding()
    dark = resolve_slot(row, SLOT_DARK)
    light = resolve_slot(row, SLOT_LIGHT)
    return {
        "name": row.get("name") or DEFAULT_NAME,
        # Relative, not absolute: the API's own address is the client's
        # `VITE_API_BASE`, and it is the client that knows it.
        "logo_url": f"/public/branding/logo/{dark}" if dark else "",
        "logo_url_light": f"/public/branding/logo/{light}" if light else "",
        # Change whenever the marks do — what an already-open tab compares.
        "stamp": dark,
        "stamp_light": light,
    }


# ===========================================================================
# Routes — PUBLIC. See the module docstring for why.
# ===========================================================================
@router.get("/public/branding")
def public_branding() -> dict:
    """The app's name and the address of its mark, for anybody at all.

    Called once per page load, before the sign-in card is drawn, and remembered
    in the browser between loads so a reload never flashes the built-in name.
    """
    return public_payload()


@router.get("/public/branding/logo/{stamp}")
def public_branding_logo(stamp: str):
    """One uploaded logo, by file id. 404 when there is none — the app draws its
    own mark instead.

    ⚠ ONE ROUTE FOR BOTH SLOTS, because the address is the FILE's id and a file
    does not care which theme asked for it. A `/logo/dark` and a `/logo/light`
    would have been two addresses whose contents change under them, which is the
    caching problem this whole scheme exists to avoid.

    ⚠ CACHED FOREVER ONLY WHEN THE STAMP IS A CURRENT ONE. A stale stamp still
    serves a live file (so a tab holding an old address is never shown a broken
    image) but tells the browser not to keep it — pinning today's bytes to
    yesterday's URL is how a logo change comes back after a hard refresh.
    """
    if not _ID_RE.match(stamp or ""):
        raise HTTPException(status_code=404, detail="No logo.")
    row = get_branding()
    live = [row.get(slot_field(s)) or "" for s in SLOTS]
    live = [i for i in live if i]
    if not live:
        raise HTTPException(status_code=404, detail="No logo.")
    # The asked-for file when it is still one of ours; otherwise whichever is
    # left, so an open tab gets a mark rather than a gap.
    serve = stamp if stamp in live else live[0]
    path = logo_path(serve)
    if not os.path.isfile(path):
        raise HTTPException(status_code=404, detail="No logo.")
    cache = "public, max-age=31536000, immutable" if serve == stamp else "no-store"
    return FileResponse(path, media_type="image/png", headers={"Cache-Control": cache})

File: client/src/branding.js
// What the app is CALLED and what its mark looks like — the browser's half of
// `server/branding.py`.
//
// THE PROBLEM THIS SOLVES: the product name was typed into eight components and
// the mark was drawn in two more, so renaming the app was a code change. It is
// an admin panel field now, and the whole point is that ONE save lands
// EVERYWHERE — the rail, the sign-in card, the landing page, the admin top bar,
// a shared storyboard link, the browser tab and the favicon — without any screen
// keeping a copy of the answer.
//
// ⚠ A MODULE-LEVEL STORE, NOT A CONTEXT, exactly like `entitlements.js` — read
// its header for the full reasoning. The short version: a Context here would
// mean a provider around the shell AND around the logged-out landing page AND
// around the public storyboard viewer, which are three different roots.
//
// ⚠ **IT IS REMEMBERED IN `localStorage`, AND THAT IS THE POINT OF THE WHOLE
// FILE.** The name is drawn in the first paint, before any request can have
// answered. Without a remembered answer every reload would show the built-in
// "Aniwala AI Studio" for a beat and then flip to the customer's name — on the
// sign-in card, which is the first thing anybody sees. Same fix, same reasoning
// as the remembered entitlements and the remembered work counts in `api.js`.
//
// ⚠ **TWO LOGOS, ONE PER THEME.** A logo is a flat picture: unlike the drawn
// mark in `Logo.jsx`, which is painted in `currentColor` and re-colours itself
// for free, an uploaded white wordmark vanishes into the light theme. The server
// resolves the fallback (one upload still covers both), so BOTH `logoUrl` and
// `logoUrlLight` always point at a real file when there is any logo at all —
// there is no "borrow the other one" rule on this side.
//
// ⚠ **AND IT FAILS BACK TO THE BUILT-IN, NEVER TO NOTHING.** A failed call keeps
// whatever is already on screen. An app that draws no title at all is a worse
// outage than one showing yesterday's name.
//
// ⚠ **AND THE COLOURS RIDE ALONG WITH THE NAME, IN ONE CALL.** The palette an
// administrator chose on the Brand screen is part of what the app is CALLED as
// far as a browser is concerned — both are public, both are needed before the
// sign-in card is drawn, and both must be remembered or the first paint is the
// wrong one. A second `/public/theme` call would be a second thing to fail, a
// second thing to remember, and a second chance for the two to disagree about
// which brand is live. The maths that turns two hex strings into every token
// the app paints with is in `palette.js`; this file only carries them.
import * as api from "./api.js";
import { applyPalette, DEFAULT_PALETTE, normalisePalette } from "./palette.js";

const STORE_KEY = "cas_brand";

// What the app ships as. ⚠ KEEP IN STEP WITH `DEFAULT_NAME` IN
// `server/branding.py` — the server falls back to its copy, this falls back to
// this one, and they are only ever both used when the two cannot talk.
export const DEFAULT_BRAND = Object.freeze({
  name: "Aniwala AI Studio",
  // "" in both means: draw the built-in mark (`components/Logo.jsx`).
  logoUrl: "",
  logoUrlLight: "",
  stamp: "",
  stampLight: "",
  // The app as it ships. ⚠ This palette injects NO CSS at all — see
  // `palette.js`. It is the absence of an override, not a copy of the
  // stylesheet.
  palette: DEFAULT_PALETTE,
});

// The favicon the app ships with, restored when the uploaded logos are removed.
const DEFAULT_ICON = { href: "/favicon.svg", type: "image/svg+xml" };

/** `{name, logo_url, logo_url_light, …}` from the wire → the shape the app uses. */
function shape(raw) {
  const name = (raw?.name || "").trim();
  // ⚠ ABSOLUTE, BUILT HERE. The server answers with PATHS because the API's own
  // address is `VITE_API_BASE` and only this side knows it. Every consumer is an
  // `<img src>` or a `<link href>`, which resolve against the PAGE — so a bare
  // path would ask Vite on :5173 for a file the API holds on :8000.
  const dark = raw?.logo_url ? api.absoluteUrl(raw.logo_url) : "";
  // ⚠ FALLS BACK TO THE DARK ONE HERE TOO, even though the server already
  // resolved it. This is what makes an answer REMEMBERED BEFORE the light slot
  // existed still work: it has no `logo_url_light` at all, and without this the
  // light theme would draw nothing for one page load after the upgrade.
  const light = raw?.logo_url_light ? api.absoluteUrl(raw.logo_url_light) : dark;
  return Object.freeze({
    name: name || DEFAULT_BRAND.name,
    logoUrl: dark,
    logoUrlLight: light,
    stamp: raw?.stamp || "",
    stampLight: raw?.stamp_light || raw?.stamp || "",
    // ⚠ `normalisePalette` NEVER THROWS AND NEVER RETURNS A PARTIAL ANSWER, and
    // that is load-bearing here: this same function shapes the REMEMBERED
    // answer, which on the release that adds colours has no `accent` in it at
    // all, and can also be whatever somebody typed into localStorage.
    palette: normalisePalette({
      id: raw?.theme_id,
      accent: raw?.accent,
      ground: raw?.ground,
    }),
  });
}

function remembered() {
  try {
    const raw = JSON.parse(localStorage.getItem(STORE_KEY) || "null");
    return raw && typeof raw === "object" ? shape(raw) : DEFAULT_BRAND;
  } catch {
    // Private mode, or somebody edited it by hand. No memory is not an error.
    return DEFAULT_BRAND;
  }
}

function remember(raw) {
  try {
    localStorage.setItem(STORE_KEY, JSON.stringify(raw));
  } catch {
    // Storage full or disabled — the app just flashes the default next reload.
  }
}

// ---------------------------------------------------------------------------
// The store
// ---------------------------------------------------------------------------
// One object, REPLACED and never mutated, so `useSyncExternalStore` can compare
// snapshots by identity. Seeded from the remembered answer at module load —
// which is before the first render, and is what removes the flash.
let _current = remembered();
const _subscribers = new Set();

/** The whole current brand, by reference. Stable between `setBrand` calls. */
export function getBrand() {
  return _current;
}

export function subscribe(fn) {
  _subscribers.add(fn);
  return () => _subscribers.delete(fn);
}

/**
 * Hand a fresh answer to every screen at once.
 *
 * ⚠ THE ADMIN PANEL CALLS THIS DIRECTLY AFTER A SAVE, and that is what makes
 * the change land instantly in the rail behind the panel instead of on the next
 * reload. The save's response is the same shape the public call returns, so
 * there is one path in and one path out.
 */
export function setBrand(raw) {
  const next = shape(raw);
  // Identity is the snapshot test, so an unchanged answer must not produce a new
  // object — a poll that re-set the same brand would otherwise re-render the
  // whole app on every tick.
  if (
    next.name === _current.name &&
    next.logoUrl === _current.logoUrl &&
    next.logoUrlLight === _current.logoUrlLight &&
    next.palette.id === _current.palette.id &&
    next.palette.accent === _current.palette.accent &&
    next.palette.ground === _current.palette.ground
  ) {
    return _current;
  }
  _current = next;
  remember({
    name: next.name,
    logo_url: raw?.logo_url || "",
    logo_url_light: raw?.logo_url_light || "",
    stamp: next.stamp,
    stamp_light: next.stampLight,
    // ⚠ REMEMBERED IN THE WIRE'S OWN FIELD NAMES, not the shaped ones, because
    // `remembered()` feeds this straight back through `shape`. A `palette`
    // object saved here would come back as a brand with no colours.
    theme_id: next.palette.id,
    accent: next.palette.accent,
    ground: next.palette.ground,
  });
  applyToDocument(next);
  for (const fn of _subscribers) fn();
  return _current;
}

/** Which logo THIS theme should draw. `dark` unless `<html>` says otherwise. */
function markFor(brand, theme) {
  return theme === "light" ? brand.logoUrlLight : brand.logoUrl;
}

/**
 * The browser tab: the title and the favicon.
 *
 * ⚠ THE TAB IS PART OF "EVERYWHERE" AND IT IS THE HALF PEOPLE FORGET. A rail
 * that says "Acme Studio" above a tab that still says "Aniwala AI Studio" is the
 * exact bug this whole change exists to remove — and the favicon is worse,
 * because a stale one survives in the bookmark.
 *
 * ⚠ THE FAVICON FOLLOWS THE THEME TOO, and it is the one place that needs JS to
 * do it: the in-app mark swaps in pure CSS off `<html data-theme>` (see
 * `base.css`), but a `<link rel=icon>` has no styling and a browser will not
 * pick between two of them. So this is re-run whenever `data-theme` changes —
 * see the observer at the foot of the file.
 *
 * The `<link>` is REPLACED, not edited: some browsers ignore an href changed in
 * place on the element they already parsed.
 */
export function applyToDocument(brand = _current) {
  if (typeof document === "undefined") return;
  document.title = brand.name;

  // ⚠ THE PALETTE IS STAMPED HERE, BESIDE THE TITLE AND THE FAVICON, AND NOT IN
  // A REACT EFFECT. Three of the screens that must be painted in the chosen
  // colours — the landing page, the sign-in card and the public storyboard
  // viewer — are outside the app shell, and one of them (`main.jsx`) runs this
  // BEFORE React mounts at all. An effect somewhere in the tree would repaint
  // the app a beat after it was drawn, which is the flash this file exists to
  // remove. Idempotent: `applyPalette` replaces one `<style>` and does nothing
  // when the CSS has not changed.
  applyPalette(brand.palette);

  const head = document.head;
  if (!head) return;
  const mark = markFor(brand, document.documentElement?.dataset?.theme);
  const icon = mark ? { href: mark, type: "image/png" } : DEFAULT_ICON;
  for (const old of head.querySelectorAll("link[rel='icon']")) old.remove();
  const link = document.createElement("link");
  link.rel = "icon";
  link.type = icon.type;
  link.href = icon.href;
  head.appendChild(link);
}

/**
 * Ask the server what the app is called. Safe to call more than once.
 *
 * ⚠ NO TOKEN AND NO ACCOUNT — the route is public because the sign-in card
 * needs the answer BEFORE anybody has signed in. See `server/branding.py`.
 *
 * ⚠ ITS FAILURE IS SILENT ON PURPOSE. There is nothing a visitor could do about
 * it and nothing to show them: the remembered name (or the built-in one) is
 * already on screen and stays there.
 */
export function loadBranding() {
  return api
    .publicBranding()
    .then((raw) => setBrand(raw))
    .catch(() => _current);
}

// ---------------------------------------------------------------------------
// Following the theme
// ---------------------------------------------------------------------------
// ⚠ AN OBSERVER, NOT A CALL FROM `theme.js`, and the reason is the number of
// places that flip the theme: `App.jsx`'s effect, the landing page's own switch,
// the admin bar's, and `main.jsx` at boot. Making each of them also re-stamp the
// favicon is four call sites and a fifth waiting to be forgotten — the same
// "everywhere" bug this file exists to end. `data-theme` on `<html>` is the ONE
// fact all four already agree on, so it is the thing to watch.
//
// Only the favicon needs this. The mark inside the app swaps in CSS.
if (typeof MutationObserver !== "undefined" && typeof document !== "undefined") {
  new MutationObserver(() => applyToDocument()).observe(document.documentElement, {
    attributes: true,
    attributeFilter: ["data-theme"],
  });
}
